test/signalrclient-e2e-tests/hub_connection_tests.cpp

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#pragma once
+
+#include "stdafx.h"
+#include <string>
+#include "cpprest\details\basic_types.h"
+#include "cpprest\json.h" 
+#include "connection.h"
+#include "hub_connection.h"
+#include "url.h"
+
+TEST(hub_connection_tests, connection_status_start_stop_start_reconnect)
+{
+    auto hubConn = std::make_shared<signalr::hub_connection>(url + U("SignalR"));
+    auto weak_hubConn = std::weak_ptr<signalr::hub_connection>(hubConn);
+    auto reconnecting_event = std::make_shared<pplx::event>();
+    auto reconnected_event = std::make_shared<pplx::event>();
+
+    auto hubProxy = hubConn->create_hub_proxy(U("hubConnection"));
+
+    hubConn->set_reconnecting([weak_hubConn, reconnecting_event]()
+    {
+        ASSERT_EQ(weak_hubConn.lock()->get_connection_state(), signalr::connection_state::reconnecting);
```

Comment:
AV. You are using `weak_connection` as a result if the actual instance will get destroyed `lock()` will return `nullptr`. You need to check against `nullptr` otherwise you would call `get_connection_state()` on `nullptr` which will AV. Instead of:

    ASSERT_EQ(weak_hubConn.lock()->get_connection_state(), signalr::connection_state::reconnecting);

you need to do:

    auto conn = weak_hub_conn.lock();
    if(conn)
    {
         ASSERT_EQ(conn->get_connection_state(), signalr::connection_state::reconnecting);
    }